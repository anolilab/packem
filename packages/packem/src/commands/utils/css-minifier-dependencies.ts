const cssMinifierDependencies = {
    cssnano: ["cssnano"],
    lightningcss: ["lightningcss"],
};

export default cssMinifierDependencies;
