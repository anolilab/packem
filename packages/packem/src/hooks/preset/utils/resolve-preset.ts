import { createJiti } from "jiti";

import type { BuildConfig, BuildPreset } from "../../../types";
import autoPreset from "../auto";

const resolvePreset = async (preset: BuildPreset | string, rootDirectory: string): Promise<BuildConfig> => {
    if (preset === "auto") {
        // eslint-disable-next-line no-param-reassign
        preset = autoPreset;
    } else if (preset === "none") {
        return {};
    } else if (typeof preset === "string") {
        // eslint-disable-next-line no-param-reassign
        preset = (await createJiti(rootDirectory).import(preset)) || {};
    }

    if (typeof preset === "function") {
        // eslint-disable-next-line no-param-reassign
        preset = preset();
    }

    return preset as BuildConfig;
};

export default resolvePreset;
