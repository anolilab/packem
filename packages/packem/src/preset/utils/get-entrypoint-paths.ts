import { normalize } from "@visulima/path";

const getEntrypointPaths = (path: string): string[] => {
    const segments = normalize(path).split("/");

    return segments.map((_, index) => segments.slice(index).join("/")).filter(Boolean);
};

export default getEntrypointPaths;
