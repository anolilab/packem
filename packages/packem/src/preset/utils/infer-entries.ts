import { existsSync } from "node:fs";
import { env } from "node:process";

import type { PackageJson } from "@visulima/package";
import { resolve } from "@visulima/path";

import { RUNTIME_EXPORT_CONVENTIONS } from "../../constants";
import type { BuildContext, BuildEntry, InferEntriesResult, Runtime } from "../../types";
import type { OutputDescriptor } from "../../utils/extract-export-filenames";
import { extractExportFilenames } from "../../utils/extract-export-filenames";
import { inferExportTypeFromFileName } from "../../utils/infer-export-type";

const getEnvironment = (output: OutputDescriptor): "production" | "development" => {
    if (output.key === "exports" && output.subKey === "production") {
        return "production";
    }

    if (output.key === "exports" && output.subKey === "development") {
        return "development";
    }

    if (env.NODE_ENV) {
        if (!env.NODE_ENV.includes("production") && !env.NODE_ENV.includes("development")) {
            throw new Error(`Invalid NODE_ENV value: ${env.NODE_ENV}, must be either "production" or "development".`);
        }

        return env.NODE_ENV as "production" | "development";
    }

    return "production";
};

const createOrUpdateEntry = (
    entries: BuildEntry[],
    input: string,
    isDirectory: boolean,
    outputSlug: string,
    output: OutputDescriptor,
    packageJson: PackageJson,
    declaration: undefined | false | true | "compatible" | "node16",
    // eslint-disable-next-line sonarjs/cognitive-complexity
): void => {
    const entry = entries.find((index) => index.input === input) ?? (entries[entries.push({ input }) - 1] as BuildEntry);

    if (isDirectory) {
        entry.outDir = outputSlug;
    }

    if (output.isExecutable) {
        entry.executable = true;

        entry.declaration = false;
        entry.cjs = packageJson.type === "commonjs";
        entry.esm = packageJson.type === "module";
    } else {
        if (/\.d\.[mc]?ts$/.test(output.file) && declaration !== false) {
            entry.declaration = declaration;
        }

        if (output.type === "cjs") {
            entry.cjs = true;
        }

        if (output.type === "esm") {
            entry.esm = true;
        }

        // eslint-disable-next-line no-loops/no-loops,no-restricted-syntax
        for (const runtime of RUNTIME_EXPORT_CONVENTIONS) {
            if (output.file.includes(runtime + ".")) {
                entry.runtime = runtime as Runtime;

                break;
            }
        }
    }

    if (entry.runtime === undefined) {
        entry.runtime = "node";
    }

    entry.environment = getEnvironment(output);
};

const ENDING_RE = /(?:\.d\.[mc]?ts|\.\w+)$/;

/**
 * Infer entries from package files.
 *
 * @param {PackageJson} packageJson
 * @param {string[]} sourceFiles A list of source files to use for inferring entries.
 * @param {BuildContext} context
 * @returns {InferEntriesResult}
 */
const inferEntries = (
    packageJson: PackageJson,
    sourceFiles: string[],
    context: BuildContext,
    // eslint-disable-next-line sonarjs/cognitive-complexity
): InferEntriesResult => {
    const warnings = [];

    // Sort files so least-nested files are first
    sourceFiles.sort((a, b) => a.split("/").length - b.split("/").length);

    // Come up with a list of all output files & their formats
    const outputs = extractExportFilenames(packageJson.exports, packageJson.type ?? "commonjs", context.options.declaration);

    if (packageJson.bin) {
        const binaries = typeof packageJson.bin === "string" ? [packageJson.bin] : Object.values(packageJson.bin);

        // eslint-disable-next-line no-loops/no-loops,no-restricted-syntax
        for (const file of binaries) {
            outputs.push({ file: file as string, isExecutable: true, key: "bin" });
        }
    }

    if (packageJson.main) {
        outputs.push({
            file: packageJson.main,
            key: "main",
            type: inferExportTypeFromFileName(packageJson.main) ?? (packageJson.type === "module" ? "esm" : "cjs"),
        });
    }

    // Defacto module entry-point for bundlers (not Node.js)
    // https://github.com/dherman/defense-of-dot-js/blob/master/proposal.md
    if (packageJson.module) {
        outputs.push({ file: packageJson.module, key: "module", type: "esm" });
    }

    // Entry point for TypeScript
    if (context.options.declaration !== false && (packageJson.types || packageJson.typings)) {
        outputs.push({ file: (packageJson.types ?? packageJson.typings) as string, key: "types" });
    }

    // Try to detect output types
    const isESMPackage = packageJson.type === "module";

    // eslint-disable-next-line no-loops/no-loops,no-restricted-syntax
    for (const output of outputs.filter((o) => !o.type)) {
        const isJS = output.file.endsWith(".js");

        if ((isESMPackage && isJS) || output.file.endsWith(".mjs")) {
            output.type = "esm";
        } else if ((!isESMPackage && isJS) || output.file.endsWith(".cjs")) {
            output.type = "cjs";
        }
    }

    // Infer entries from package files
    const entries: BuildEntry[] = [];

    // eslint-disable-next-line no-loops/no-loops,no-restricted-syntax
    for (const output of outputs) {
        // Supported output file extensions are `.d.ts`, `.cjs` and `.mjs`
        // But we support any file extension here in case user has extended rollup options
        const outputSlug = output.file.replace(/(?:\*[^/\\]|\.d\.[mc]?ts|\.\w+)$/, "");

        const isDirectory = outputSlug.endsWith("/");

        // Skip top level directory
        if (isDirectory && ["./", "/"].includes(outputSlug)) {
            // eslint-disable-next-line no-continue
            continue;
        }

        // eslint-disable-next-line @rushstack/security/no-unsafe-regexp,security/detect-non-literal-regexp
        const sourceSlug = outputSlug.replace(new RegExp("(./)?" + context.options.outDir), context.options.sourceDir);

        // @see https://nodejs.org/docs/latest-v16.x/api/packages.html#subpath-patterns
        if (output.file.includes("/*") && output.key === "exports") {
            context.logger.warn("Private subfolders are not supported, if you need this feature please open an issue on GitHub.");

            const inputs: string[] = [];

            // eslint-disable-next-line @rushstack/security/no-unsafe-regexp,security/detect-non-literal-regexp
            const SOURCE_RE = new RegExp("(?<=/|$)" + sourceSlug.replace("*", "(.*)") + (isDirectory ? "" : "\\.\\w+"));
            // eslint-disable-next-line no-loops/no-loops,no-restricted-syntax
            for (const source of sourceFiles) {
                if (SOURCE_RE.test(source)) {
                    inputs.push(source.replace(ENDING_RE, ""));
                }
            }

            if (inputs.length === 0) {
                warnings.push(`Could not find entrypoints for \`${output.file}\``);

                // eslint-disable-next-line no-continue
                continue;
            }

            // eslint-disable-next-line no-loops/no-loops,no-restricted-syntax
            for (const input of inputs) {
                createOrUpdateEntry(entries, input, isDirectory, outputSlug, output, packageJson, context.options.declaration);
            }

            // eslint-disable-next-line no-continue
            continue;
        }

        // eslint-disable-next-line @rushstack/security/no-unsafe-regexp,security/detect-non-literal-regexp
        const SOURCE_RE = new RegExp("(?<=/|$)" + sourceSlug + (isDirectory ? "" : "\\.\\w+"));

        const input = sourceFiles.find((index) => SOURCE_RE.test(index))?.replace(ENDING_RE, "");

        if (!input) {
            // eslint-disable-next-line security/detect-non-literal-fs-filename
            if (!existsSync(resolve(context.options.rootDir, output.file))) {
                warnings.push(`Could not find entrypoint for \`${output.file}\``);
            }

            // eslint-disable-next-line no-continue
            continue;
        }

        createOrUpdateEntry(entries, input, isDirectory, outputSlug, output, packageJson, context.options.declaration);
    }

    return { entries, warnings };
};

export default inferEntries;
