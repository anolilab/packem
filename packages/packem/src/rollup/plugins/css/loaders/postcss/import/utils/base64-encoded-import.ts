import type { Condition } from "../types";
import formatImportPrelude from "./format-import-prelude";

const base64EncodedConditionalImport = (prelude: string, conditions: Condition[]): string => {
    conditions.reverse();

    const first = conditions.pop() as Condition;

    let parameters = `${prelude} ${formatImportPrelude(first.layer, first.media, first.supports)}`;

    for (const condition of conditions) {
        parameters = `'data:text/css;base64,${Buffer.from(`@import ${parameters}`).toString("base64")}' ${formatImportPrelude(
            condition.layer,
            condition.media,
            condition.supports,
        )}`;
    }

    return parameters;
};

export default base64EncodedConditionalImport;
