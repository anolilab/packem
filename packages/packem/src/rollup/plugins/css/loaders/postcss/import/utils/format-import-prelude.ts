const formatImportPrelude = (layer: string | undefined, media: string | undefined, supports: string | undefined): string => {
    const parts = [];

    if (layer !== undefined) {
        let layerParameters = "layer";

        if (layer) {
            layerParameters = `layer(${layer})`;
        }

        parts.push(layerParameters);
    }

    if (supports !== undefined) {
        parts.push(`supports(${supports})`);
    }

    if (media !== undefined) {
        parts.push(media);
    }

    return parts.join(" ");
};

export default formatImportPrelude;
