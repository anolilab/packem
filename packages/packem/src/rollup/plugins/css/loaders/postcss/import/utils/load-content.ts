import { readFileSync } from "@visulima/fs";

import { contents, isValid } from "./data-url";

const loadContent = (filename: string): string => {
    if (isValid(filename)) {
        return contents(filename);
    }

    return readFileSync(filename);
};

export default loadContent;
