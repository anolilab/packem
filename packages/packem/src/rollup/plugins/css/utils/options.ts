import type { Result } from "postcss-load-config";

import type { LoaderContext } from "../loaders/types";
import type { InternalStyleOptions,StyleOptions } from "../types";
import arrayFmt from "./array-fmt";
import loadModule from "./load-module";

interface Mode {
    emit: InternalStyleOptions["emit"];
    extract: InternalStyleOptions["extract"];
    inject: InternalStyleOptions["inject"];
}

const modes = ["inject", "extract", "emit"];
const modesFmt = arrayFmt(modes);

type PCSSOption = "parser" | "plugin" | "stringifier" | "syntax";

export const inferModeOption = (mode: StyleOptions["mode"]): Mode => {
    const m = Array.isArray(mode) ? mode : ([mode] as const);

    if (m[0] && !modes.includes(m[0])) {
        throw new Error(`Incorrect mode provided, allowed modes are ${modesFmt}`);
    }

    return {
        emit: m[0] === "emit",
        extract: m[0] === "extract" && (m[1] ?? true),
        inject: (!m[0] || m[0] === "inject") && (m[1] ?? true),
    };
}

export const inferSourceMapOption = (sourceMap: StyleOptions["sourceMap"]): LoaderContext["sourceMap"] => {
    const sm = Array.isArray(sourceMap) ? sourceMap : ([sourceMap] as const);

    if (!sm[0]) {
        return false;
    }

    return { content: true, ...sm[1], inline: sm[0] === "inline" };
};

export const inferHandlerOption = <T extends { alias?: Record<string, string> }>(option: T | boolean | undefined, alias: T["alias"]): T | false => {
    const opt = inferOption(option, {} as T);

    if (alias && typeof opt === "object" && !opt.alias) {
        opt.alias = alias;
    }

    return opt;
};

export const ensurePCSSOption = <T>(option: T | string, type: PCSSOption): T => {
    if (typeof option !== "string") {
        return option;
    }

    const module = loadModule(option);

    if (!module) {
        throw new Error(`Unable to load PostCSS ${type} \`${option}\``);
    }

    return module as T;
};

export const ensurePCSSPlugins = (plugins: StyleOptions["postcss"]["plugins"]): Result["plugins"] => {
    if (plugins === undefined) {
        return [];
    }

    if (typeof plugins !== "object") {
        throw new TypeError("`plugins` option must be an array or an object!");
    }

    const ps: Result["plugins"] = [];

    for (const p of Array.isArray(plugins) ? plugins : Object.entries(plugins)) {
        if (!p) {
            // eslint-disable-next-line no-continue
            continue;
        }

        if (!Array.isArray(p)) {
            ps.push(ensurePCSSOption(p, "plugin"));

            continue;
        }

        const [plug, options] = p;

        if (options) {
            ps.push(ensurePCSSOption(plug, "plugin")(options));
        } else {
            ps.push(ensurePCSSOption(plug, "plugin"));
        }
    }

    return ps;
};

export const inferOption = <T, TDefine extends T | boolean>(option: T | boolean | undefined, defaultValue: TDefine): T | TDefine | false => {
    if (typeof option === "boolean") {
        return option && ({} as TDefine);
    }

    if (typeof option === "object") {
        return option;
    }

    return defaultValue;
}
