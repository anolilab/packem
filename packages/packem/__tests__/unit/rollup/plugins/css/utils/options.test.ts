import { describe, expect, it } from "vitest";

import { ensurePCSSOption } from "../../../../../../src/rollup/plugins/css/utils/options";

describe("option", () => {
    it("wrong postcss option", () => {
        expect.assertions(1);

        expect(() => ensurePCSSOption("pumpinizer", "plugin")).toThrowErrorMatchingSnapshot();
    });
});
