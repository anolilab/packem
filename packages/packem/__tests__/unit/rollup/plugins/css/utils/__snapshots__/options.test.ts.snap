// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`option > wrong postcss option 1`] = `[Error: Unable to load PostCSS plugin \`pumpinizer\`]`;
